[text_input] prepare for custom text input sources

jpnurmi opened the pull request:
## Description

The existing `TextInputConnection` class has been made abstract and method channel-agnostic. There is now a separate internal class, `_TextInputChannelConnection`, that implements the default type of text input connection that communicates with platform plugins via the text input system channel.

The main motivation of the separation is to allow custom text input sources that can replace the method channel-based implementation and for example, communicate with an in-app virtual keyboard on (embedded) platforms that do not have a system-provided keyboard.

To be able to eliminate direct method channel calls from the `TextInput` class, the following new methods have been introduced:
- `TextInput.detach()`
- `TextInput.reset()`
- `TextInputConnection.hide()`
- `TextInputConnection.setClient()`
- `TextInputConnection.clearClient()`

Consequently, the following methods have been deprecated:
- `TextInputConnection.close()` in favor of `TextInput.detach()`
- `TextInputConnection.connectionClosedReceived()` in favor of `TextInput.reset()`

## Related PRs

This PR was split off from #69146.

## Related Issues

#68988

## Tests

I added the following tests:
- `test/services/text_input_test.dart`:
  - _TextInput message channels_:
    - _text input client is requested to hide on detach_
    - _old client is detached when a new client is attached_
    - _text input connection is reset_

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jpnurmi committed:
[text_input] prepare for custom text input sources

The existing TextInputConnection class has been made abstract and
method channel-agnostic. There is now a separate internal class,
_TextInputChannelConnection, that implements the default type of text
input connection that communicates with platform plugins via the text
input system channel.

The main motivation of the separation is to allow custom text input
sources that can replace the method channel-based implementation and
for example, communicate with an in-app virtual keyboard on (embedded)
platforms that do not have a keyboard provided by the system

Ref: #68988

jpnurmi commented:
Rebased again in hopes to pass the CI (unrelated git fetch failures on macOS).

justinmc reviewed (commented):
Thanks for splitting this out from https://github.com/flutter/flutter/pull/69146, should be easier to review and merge.

It looks like this could hypothetically break someone that was subclassing TextInputConnection.  Do we know if anyone might be doing that?  I've submitted this PR to have the internal Google tests run on it, and if it passes there then I'm more confident that it's not a problem.

CC @LongCatIsLooong since you reviewed https://github.com/flutter/flutter/pull/69146.

jpnurmi committed:
Merge remote-tracking branch 'origin/master' into text-input-channel-connection

jpnurmi committed:
[docs] fix spelling and deprecation version

jpnurmi commented:
Thanks for the review!

> It looks like this could hypothetically break someone that was subclassing TextInputConnection. Do we know if anyone might be doing that? I've submitted this PR to have the internal Google tests run on it, and if it passes there then I'm more confident that it's not a problem.

I believe it's not a breaking change because the constructor was private before, so it wasn't possible to create a `TextInputConnection` instance from the outside of `text_input.dart`.

justinmc commented:
Ah I missed that, you're right!  Should be fine then.  I'm on board once all the tests pass.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -938,11 +946,121 @@ class TextInputConnection {
+    required TextAlign textAlign,
+  });
+
+  /// Use [TextInput.detach] instead.
```
nit: keep the old documentation, the deprecation notice should be enough for saying this is replaced by `detach`.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1130,7 +1288,7 @@ RawFloatingCursorPoint _toTextPoint(FloatingCursorDragState state, Map<String, d
   TextInput._() {
     _channel = SystemChannels.textInput;
-    _channel.setMethodCallHandler(_handleTextInputInvocation);
+    _channel.setMethodCallHandler(_TextInputChannelConnection._handleTextInputInvocation);
```
any chance we can set the handler when we attach to a new connection and unset when we detach from this connection? It sounds wrong that when we switch connection the old handler is still listening to the text input channel

jpnurmi replied:
I recall we had it like that in the beginning, but there was a problem with how `_DefaultBinaryMessenger` deal with unhandled messages: https://github.com/flutter/flutter/pull/69146#discussion_r523240713

LongCatIsLooong replied:
Ah right thanks for pointing that out. It feels like we should let `TextInputConnection` own the text input channel (at least pretend to own it since it's a singleton), and decide what to do when it disconnects? Now that `TextInputConnection` handles the communication between the client and the keyboard,  should `TextInput` just be responsible for creating/switching connections (attach, detach)? It's OK to leave `finishAutofillContext` in `TextInput` for now since we can still access the method channel via `SystemChannels.textInput`. WDYT?

jpnurmi replied:
We are still missing a good place for static `TextInputConnection` init and cleanup. The upcoming `TextInputSource` could help with that because it could have such methods that init and cleanup things that apply to all specific type of input connection instances.

The default `TextInputSource` implementation could act as the "owner" of the input channel and pass it over to input connections. It could also set a stub method call handler that discards all incoming messages when not in use.

> It's OK to leave finishAutofillContext in TextInput for now since we can still access the method channel via SystemChannels.textInput.

Accessing `SystemChannels.textInput` directly would unfortunately break autofill tests that rely on `TextInput.setChannel()`. We could actually solve this problem too with `TextInputSource`, though. Since `finishAutofillContext()` is not tied to any specific input connection, we could redirect it to the current input source.

jpnurmi replied:
I added an internal `_TextInputSourrce` to address these issues. I'll follow up with the rest in #69146.

jpnurmi committed:
Merge remote-tracking branch 'origin/master' into text-input-channel-connection

jpnurmi committed:
[docs] restore original docs for deprecated methods

justinmc reviewed (approved):
LGTM but I'll wait for @LongCatIsLooong's final approval.

jpnurmi committed:
Merge remote-tracking branch 'origin/master' into text-input-channel-connection

jpnurmi committed:
[text_input] eliminate remaining MethodChannel references in TextInput

Later on, _TextInputSource will be made public, abstract and documented
so that it can be used as a base class for custom input sources.

jpnurmi committed:
Remove unused TextInput._setSource() for now to pass analysis

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1420,9 +1492,42 @@ class TextInput {
+  static MethodChannel? _channel;
+
+  static void setChannel(MethodChannel newChannel) {
+    _channel = newChannel..setMethodCallHandler(_handleTextInputInvocation);
```
should we call clean up on the previous channel?

jpnurmi replied:
This is only used in tests for setting up a testable environment via `@visibleForTesting TextInput.setChannel()`. This will be moved to the upcoming private default implementation of `TextInputSource` so it won't be visible to the outside.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1231,80 +1384,13 @@ class TextInput {
 
-  late MethodChannel _channel;
-
+  final _TextInputSource _currentSource = _TextInputSource();
```
This class will be made public in a different PR?

jpnurmi replied:
The next step is to make this an abstract public class and move the default implementation into a private subclass. We have most of it already covered in the original PR #69146 but especially the docs need some work. I will rebase #69146 and continue where we left off.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1315,59 +1401,45 @@ class TextInput {
+  ///
+  /// This function should be called to reset the current text input connection
+  /// in case the platform sent a notification informing the connection is
+  /// closed.
```
nit: maybe crossreference `connectionClosed` in the documentation.

LongCatIsLooong reviewed (approved):
LGTM. Thank you!

jpnurmi committed:
[docs] link TextInput.reset to TextInputClient.connectionClosed

jpnurmi commented:
Is the Google testing failure somehow related to this?

LongCatIsLooong commented:
Doesn't seem to be the case. I have restarted the test.

LongCatIsLooong commented:
The failure seems to be caused by something irrelevant. Could you rebase the branch to the latest commit and see if the failure goes away?

jpnurmi committed:
Merge remote-tracking branch 'origin/master' into text-input-channel-connection

jpnurmi commented:
Unrelated Mac build_tests failure. Let's try again.
